cranelift/docs/testing.md

```diff
@@ -367,3 +367,88 @@ Example:
     }
     ; run
 ```
+
+#### Environment directives
+
+Some tests need additional resources to be provided by the filetest infrastructure.
+
+When any of the following directives is present the first argument of the function is *required* to be a `i64 vmctx`.
+The filetest infrastructure will then pass a pointer to the environment struct via this argument.
+
+The environment struct is essentially a list of pointers with info about the resources requested by the directives.
+
+Currently, we only support requesting heaps, however this is a generic mechanism that should
+be able to introduce any sort of environment support that we may need later. (e.g. tables, global values, external functions)
+
+##### `heap` directive
+
+The `heap` directive allows a test to request a heap to be allocated and passed to the test via the environment struct.
+
+
+A sample heap annotation is the following:
+```
+; heap: static, size=0x1000, ptr=vmctx+0, bound=vmctx+8
+```
+
+This indicates the following:
+* `static`: We have requested a non resizabe and non-movable static heap.
+* `size=0x1000`: It has to have a size of 4096 bytes.
+* `ptr=vmctx+0`: The pointer to the address to the start of this heap is placed at offset 0 in the `vmctx` struct
+* `bound=vmctx+8`: The pointer to the address to the end of this heap is placed at offset 8 in the `vmctx` struct
+
+The `ptr` and `bound` arguments make explicit the placement of the pointers to the start and end of the heap memory in
+the environment struct. `vmctx+0` means that at offset 0 of the environment struct there will be the pointer to the start
+similarly, at offset 8 the pointer to the end.
+
+
+You can combine multiple heap annotations, in which case, their pointers are laid out sequentially in memory.
+
+```
+; heap: static, size=0x1000, ptr=vmctx+0, bound=vmctx+8
+; heap: dynamic, size=0x1000, ptr=vmctx+16, bound=vmctx+24
```

Comment:
I think the heaps may be laid out sequentially in memory but I'm not sure about the pointers: if the user writes the following,
```
; heap: static, size=0x1000, ptr=vmctx+0, bound=vmctx+16
; heap: dynamic, size=0x1000, ptr=vmctx+8, bound=vmctx+24
```
Then do bad things happen?

Reply:
Perhaps this can't be done... enforced by the parser? Maybe the docs should say something about "must be laid out sequentially..."

Reply:
If the layout is different from what we expect, we throw an error.

https://github.com/bytecodealliance/wasmtime/pull/3154/files#diff-d1f760ce6f086af21cead6aa563293fed30f716318faa1250974c60cb62058f4R22

I don't like that this interface gives the user the impression that they can change the offsets. But I think its better than being implicit.

> Maybe the docs should say something about "must be laid out sequentially..."

Yeah, I'll update the docs.